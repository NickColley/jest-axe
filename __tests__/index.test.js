const { configureAxe, axe, toHaveNoViolations } = require("../index.js");

describe("jest-axe", () => {
  describe("axe", () => {
    const failingHtmlExample = `
     <html>
       <body>
         <a href="#"></a>
       </body>
     </html>
    `;

    const failingExtendedHtmlExample = `
     <html>
       <body>
         <a href="#"></a>
         <img src="#"/>
       </body>
     </html>
    `;

    const goodHtmlExample = `
     <html>
       <body>
        <main>
         <a href="http://gov.uk">Visit GOV.UK</a>
        </main>
       </body>
     </html>
    `;

    const linkNameAxe = configureAxe({
      rules: {
        "link-name": { enabled: false },
      },
    });

    it("can be configured for global configs", async () => {
      const results = await linkNameAxe(failingHtmlExample);
      expect(results.violations).toEqual([]);
    });

    it("can pass in merged configurations to configured axe", async () => {
      const results = await linkNameAxe(failingExtendedHtmlExample, {
        rules: {
          "image-alt": { enabled: false },
          region: { enabled: false },
        },
      });
      expect(results.violations).toEqual([]);
    });

    it("returns an axe results object", async () => {
      const results = await axe(failingHtmlExample);
      expect(typeof results).toBe("object");
      expect(typeof results.violations).toBe("object");
    });

    it("should not mutate the content of document.body permanently", async () => {
      const el = document.body.appendChild(document.createElement("div"));
      await axe(goodHtmlExample);
      expect(document.body.childElementCount).toBe(1);
      expect(document.body.firstChild).toEqual(el);
    });

    it("returns violations for failing html example", async () => {
      const results = await axe(failingHtmlExample);
      const violation = results.violations[0];
      expect(violation.id).toBe("link-name");
      expect(violation.description).toBe("Ensures links have discernible text");
    });

    it("can ignore allowed failures", async () => {
      const results = await axe(failingHtmlExample, {
        rules: {
          "link-name": { enabled: false },
        },
      });
      expect(results.violations).toEqual([]);
    });

    it("returns no violations for a good html example", async () => {
      const results = await axe(goodHtmlExample);
      expect(results.violations).toEqual([]);
    });

    it("throws if input is not a string, vue element, react element, or react testing library render", () => {
      expect(() => {
        axe({});
      }).toThrow("html parameter should be an HTML string or an HTML element");
    });

    it("throws with non-html input", () => {
      expect(() => {
        axe("Hello, World");
      }).toThrow('html parameter ("Hello, World") has no elements');
    });

    it("should not mutate previous options", async () => {
      let results = await axe(failingHtmlExample, {
        rules: {
          "link-name": { enabled: false },
        },
      });
      expect(results.violations).toEqual([]);

      const configuredAxe = configureAxe({
        rules: {
          "link-name": { enabled: false },
        },
      });

      results = await configuredAxe(failingHtmlExample, {
        rules: {
          "link-name": { enabled: false },
        },
      });
      expect(results.violations).toEqual([]);

      results = await axe(failingHtmlExample);
      const violation = results.violations[0];
      expect(violation.id).toBe("link-name");
      expect(violation.description).toBe("Ensures links have discernible text");
    });
  });

  describe("toHaveNoViolations", () => {
    const failingAxeResults = {
      violations: [
        {
          id: "image-alt",
          impact: "critical",
          tags: [
            "cat.text-alternatives",
            "wcag2a",
            "wcag111",
            "section508",
            "section508.22.a",
          ],
          description:
            "Ensures <img> elements have alternate text or a role of none or presentation",
          help: "Images must have alternate text",
          helpUrl:
            "https://dequeuniversity.com/rules/axe/2.6/image-alt?application=axeAPI",
          nodes: [
            {
              any: [
                {
                  id: "has-alt",
                  data: null,
                  relatedNodes: [],
                  impact: "critical",
                  message: "Element does not have an alt attribute",
                },
                {
                  id: "aria-label",
                  data: null,
                  relatedNodes: [],
                  impact: "serious",
                  message: "aria-label attribute does not exist or is empty",
                },
                {
                  id: "aria-labelledby",
                  data: null,
                  relatedNodes: [],
                  impact: "serious",
                  message:
                    "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty or not visible",
                },
                {
                  id: "non-empty-title",
                  data: null,
                  relatedNodes: [],
                  impact: "serious",
                  message:
                    "Element has no title attribute or the title attribute is empty",
                },
                {
                  id: "role-presentation",
                  data: null,
                  relatedNodes: [],
                  impact: "minor",
                  message:
                    'Element\'s default semantics were not overridden with role="presentation"',
                },
                {
                  id: "role-none",
                  data: null,
                  relatedNodes: [],
                  impact: "minor",
                  message:
                    'Element\'s default semantics were not overridden with role="none"',
                },
              ],
              all: [],
              none: [],
              impact: "critical",
              html: '<img src="">',
              target: ["body > img"],
              failureSummary:
                'Fix any of the following:\n  Element does not have an alt attribute\n  aria-label attribute does not exist or is empty\n  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty or not visible\n  Element has no title attribute or the title attribute is empty\n  Element\'s default semantics were not overridden with role="presentation"\n  Element\'s default semantics were not overridden with role="none"',
            },
          ],
        },
      ],
    };

    const successfulAxeResults = {
      violations: [],
    };
    it("returns a jest matcher object with object", () => {
      const matcherFunction = toHaveNoViolations.toHaveNoViolations;
      expect(matcherFunction).toBeDefined();
      expect(typeof matcherFunction).toBe("function");
    });

    it("throws error if non axe results object is passed", () => {
      const matcherFunction = toHaveNoViolations.toHaveNoViolations;
      expect(() => {
        matcherFunction({});
      }).toThrow(
        "Unexpected aXe results object. No violations property found.\nDid you change the `reporter` in your aXe configuration?"
      );
    });

    it("returns pass as true when no violations are present", () => {
      const matcherFunction = toHaveNoViolations.toHaveNoViolations;
      const matcherOutput = matcherFunction(successfulAxeResults);
      expect(matcherOutput.pass).toBe(true);
    });

    it("returns same violations that are passed in the results object", () => {
      const matcherFunction = toHaveNoViolations.toHaveNoViolations;
      const matcherOutput = matcherFunction(failingAxeResults);
      expect(matcherOutput.actual).toBe(failingAxeResults.violations);
    });

    it("returns correctly formatted message when violations are present", () => {
      const matcherFunction = toHaveNoViolations.toHaveNoViolations;
      const matcherOutput = matcherFunction(failingAxeResults);
      expect(typeof matcherOutput.message).toBe("function");
      expect(matcherOutput.message()).toMatchSnapshot();
    });

    it("returns pass as false when violations are present", () => {
      const matcherFunction = toHaveNoViolations.toHaveNoViolations;
      const matcherOutput = matcherFunction(failingAxeResults);
      expect(matcherOutput.pass).toBe(false);
    });

    it("returns properly formatted text with more complex example", async () => {
      const complexHtmlExample = `
        <html>
          <body>
            <a href="#link-name"></a>
            <a href="#link-name-2"></a>
            <img src="http://example.com"/>
            <img src="http://example.com/2"/>

            <input type="text"/>
          </body>
        </html>
      `;
      const results = await axe(complexHtmlExample);
      const matcherFunction = toHaveNoViolations.toHaveNoViolations;
      const matcherOutput = matcherFunction(results);
      expect(matcherOutput.message()).toMatchSnapshot();
    });
  });
  describe("readme", () => {
    describe("first readme example", () => {
      expect.extend(toHaveNoViolations);

      it("should demonstrate this matcher`s usage", async () => {
        const render = () => '<img src="#"/>';

        // pass anything that outputs html to axe
        const html = render();

        const results = await axe(html);

        expect(() => {
          expect(results).toHaveNoViolations();
        }).toThrowErrorMatchingSnapshot();
      });
    });
    describe("readme axe config example", () => {
      expect.extend(toHaveNoViolations);

      it("should demonstrate this matcher`s usage with a custom config", async () => {
        const render = () => `
          <div>
            <img src="#"/>
          </div>
        `;

        // pass anything that outputs html to axe
        const html = render();

        const results = await axe(html, {
          rules: {
            // for demonstration only, don't disable rules that need fixing.
            "image-alt": { enabled: false },
            region: { enabled: false },
          },
        });

        expect(results).toHaveNoViolations();
      });
    });
    describe("readme axe global config example", () => {
      // Global helper file (axe-helper.js)

      const configuredAxe = configureAxe({
        rules: {
          // for demonstration only, don't disable rules that need fixing.
          "image-alt": { enabled: false },
          region: { enabled: false },
        },
      });

      const exportedAxe = configuredAxe;

      // Individual test file (test.js)
      const axe = exportedAxe; // require('./axe-helper.js')

      expect.extend(toHaveNoViolations);

      it("should demonstrate this matcher`s usage with a default config", async () => {
        const render = () => `
          <div>
            <img src="#"/>
          </div>
        `;

        // pass anything that outputs html to axe
        const html = render();

        expect(await axe(html)).toHaveNoViolations();
      });
    });
    describe("configure custom rule", () => {
      expect.extend(toHaveNoViolations);

      it("should report custom rules", async () => {
        const check = {
          id: "demo-has-data",
          evaluate(node) {
            return node.hasAttribute("data-demo-rule");
          },
          metadata: {
            impact: "serious",
            messages: {
              fail: "Error!",
            },
          },
        };
        const rule = {
          id: "demo-rule",
          selector: ".demo",
          enabled: false,
          tags: ["demo-rules"],
          any: ["demo-has-data"],
          metadata: {
            description: "Demo check",
            help: "Demo check",
          },
        };

        const configuredAxe = configureAxe({
          globalOptions: {
            rules: [rule],
            checks: [check],
          },
          rules: {
            "demo-rule": { enabled: true },
          },
        });

        const html = `
        <html>
          <body>
            <main>
              <div class="demo">
            </main>
          </body>
        </html>
        `;

        const results = await configuredAxe(html);
        expect(results.violations[0].id).toBe("demo-rule");
      });
    });

    describe("custom configuration for user impact", () => {
      const axe = configureAxe({
        // How serious the violation is. Can be one of "minor", "moderate", "serious", or "critical".
        impactLevels: ["critical"],
      });

      expect.extend(toHaveNoViolations);

      it("should pass the test, because only critical violations are noted.", async () => {
        // 1 x moderate violation -> https://dequeuniversity.com/rules/axe/4.0/region?application=axeAPI
        const render = () => `
          <div>
            <div>
              <span> some content</span>
            </div>
          </div>
        `;

        // pass anything that outputs html to axe
        const html = render();

        expect(await axe(html)).toHaveNoViolations();
      });
    });
  });
});
