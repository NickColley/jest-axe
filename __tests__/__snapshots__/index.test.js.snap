// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-axe readme first readme example should demonstrate this matcher\`s usage 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toHaveNoViolations([22m[32mexpected[39m[2m)[22m

Expected the HTML found at $('img') to have no violations:

[90m<img src="#">[39m

Received:

[31m"Images must have alternate text (image-alt)"[39m

[33mFix any of the following:[39m
[33m  Element does not have an alt attribute[39m
[33m  aria-label attribute does not exist or is empty[39m
[33m  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty[39m
[33m  Element has no title attribute[39m
[33m  Element's default semantics were not overridden with role="none" or role="presentation"[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/4.7/image-alt?application=axeAPI[39m

â”€â”€â”€â”€â”€â”€â”€â”€

Expected the HTML found at $('img') to have no violations:

[90m<img src="#">[39m

Received:

[31m"All page content should be contained by landmarks (region)"[39m

[33mFix any of the following:[39m
[33m  Some page content is not contained by landmarks[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/4.7/region?application=axeAPI[39m"
`;

exports[`jest-axe toHaveNoViolations returns correctly formatted message when violations are present 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toHaveNoViolations([22m[32mexpected[39m[2m)[22m

Expected the HTML found at $('body > img') to have no violations:

[90m<img src="">[39m

Received:

[31m"Images must have alternate text (image-alt)"[39m

[33mFix any of the following:[39m
[33m  Element does not have an alt attribute[39m
[33m  aria-label attribute does not exist or is empty[39m
[33m  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty or not visible[39m
[33m  Element has no title attribute or the title attribute is empty[39m
[33m  Element's default semantics were not overridden with role="presentation"[39m
[33m  Element's default semantics were not overridden with role="none"[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/2.6/image-alt?application=axeAPI[39m"
`;

exports[`jest-axe toHaveNoViolations returns properly formatted text with more complex example 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toHaveNoViolations([22m[32mexpected[39m[2m)[22m

Expected the HTML found at $('img[src$="example.com"]') to have no violations:

[90m<img src="http://example.com">[39m

Received:

[31m"Images must have alternate text (image-alt)"[39m

[33mFix any of the following:[39m
[33m  Element does not have an alt attribute[39m
[33m  aria-label attribute does not exist or is empty[39m
[33m  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty[39m
[33m  Element has no title attribute[39m
[33m  Element's default semantics were not overridden with role="none" or role="presentation"[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/4.7/image-alt?application=axeAPI[39m

Expected the HTML found at $('img[src="http://example.com/2"]') to have no violations:

[90m<img src="http://example.com/2">[39m

Received:

[31m"Images must have alternate text (image-alt)"[39m

[33mFix any of the following:[39m
[33m  Element does not have an alt attribute[39m
[33m  aria-label attribute does not exist or is empty[39m
[33m  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty[39m
[33m  Element has no title attribute[39m
[33m  Element's default semantics were not overridden with role="none" or role="presentation"[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/4.7/image-alt?application=axeAPI[39m

â”€â”€â”€â”€â”€â”€â”€â”€

Expected the HTML found at $('input') to have no violations:

[90m<input type="text">[39m

Received:

[31m"Form elements must have labels (label)"[39m

[33mFix any of the following:[39m
[33m  Form element does not have an implicit (wrapped) <label>[39m
[33m  Form element does not have an explicit <label>[39m
[33m  aria-label attribute does not exist or is empty[39m
[33m  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty[39m
[33m  Element has no title attribute[39m
[33m  Element has no placeholder attribute[39m
[33m  Element's default semantics were not overridden with role="none" or role="presentation"[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/4.7/label?application=axeAPI[39m

â”€â”€â”€â”€â”€â”€â”€â”€

Expected the HTML found at $('a[href$="#link-name"]') to have no violations:

[90m<a href="#link-name"></a>[39m

Received:

[31m"Links must have discernible text (link-name)"[39m

[33mFix all of the following:[39m
[33m  Element is in tab order and does not have accessible text[39m
[33m[39m
[33mFix any of the following:[39m
[33m  Element does not have text that is visible to screen readers[39m
[33m  aria-label attribute does not exist or is empty[39m
[33m  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty[39m
[33m  Element has no title attribute[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/4.7/link-name?application=axeAPI[39m

Expected the HTML found at $('a[href$="#link-name-2"]') to have no violations:

[90m<a href="#link-name-2"></a>[39m

Received:

[31m"Links must have discernible text (link-name)"[39m

[33mFix all of the following:[39m
[33m  Element is in tab order and does not have accessible text[39m
[33m[39m
[33mFix any of the following:[39m
[33m  Element does not have text that is visible to screen readers[39m
[33m  aria-label attribute does not exist or is empty[39m
[33m  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty[39m
[33m  Element has no title attribute[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/4.7/link-name?application=axeAPI[39m

â”€â”€â”€â”€â”€â”€â”€â”€

Expected the HTML found at $('img[src$="example.com"]') to have no violations:

[90m<img src="http://example.com">[39m

Received:

[31m"All page content should be contained by landmarks (region)"[39m

[33mFix any of the following:[39m
[33m  Some page content is not contained by landmarks[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/4.7/region?application=axeAPI[39m

Expected the HTML found at $('img[src="http://example.com/2"]') to have no violations:

[90m<img src="http://example.com/2">[39m

Received:

[31m"All page content should be contained by landmarks (region)"[39m

[33mFix any of the following:[39m
[33m  Some page content is not contained by landmarks[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/4.7/region?application=axeAPI[39m

Expected the HTML found at $('input') to have no violations:

[90m<input type="text">[39m

Received:

[31m"All page content should be contained by landmarks (region)"[39m

[33mFix any of the following:[39m
[33m  Some page content is not contained by landmarks[39m

You can find more information on this issue here: 
[34mhttps://dequeuniversity.com/rules/axe/4.7/region?application=axeAPI[39m"
`;
